File: soma-ui/flutter/example/lib/screens/icons_page.dart
import 'package:flutter/material.dart';
import 'package:soma_ui/soma_ui.dart';

// Curated list of ~120 common Lucide icons grouped by concept.
const _icons = <(String, IconData)>[
  // Arrows & Chevrons
  ('arrowUp', LucideIcons.arrowUp),
  ('arrowDown', LucideIcons.arrowDown),
  ('arrowLeft', LucideIcons.arrowLeft),
  ('arrowRight', LucideIcons.arrowRight),
  ('arrowUpRight', LucideIcons.arrowUpRight),
  ('arrowDownLeft', LucideIcons.arrowDownLeft),
  ('chevronUp', LucideIcons.chevronUp),
  ('chevronDown', LucideIcons.chevronDown),
  ('chevronLeft', LucideIcons.chevronLeft),
  ('chevronRight', LucideIcons.chevronRight),
  ('chevronsUpDown', LucideIcons.chevronsUpDown),
  ('chevronsLeftRight', LucideIcons.chevronsLeftRight),
  // Actions
  ('plus', LucideIcons.plus),
  ('minus', LucideIcons.minus),
  ('x', LucideIcons.x),
  ('check', LucideIcons.check),
  ('checkCheck', LucideIcons.checkCheck),
  ('copy', LucideIcons.copy),
  ('edit', LucideIcons.edit),
  ('pencil', LucideIcons.pencil),
  ('trash', LucideIcons.trash),
  ('save', LucideIcons.save),
  ('download', LucideIcons.download),
  ('upload', LucideIcons.upload),
  ('share', LucideIcons.share),
  ('send', LucideIcons.send),
  ('refreshCw', LucideIcons.refreshCw),
  ('rotateCw', LucideIcons.rotateCw),
  ('undo', LucideIcons.undo),
  ('zoomIn', LucideIcons.zoomIn),
  ('zoomOut', LucideIcons.zoomOut),
  ('expand', LucideIcons.expand),
  ('maximize', LucideIcons.maximize),
  ('minimize', LucideIcons.minimize),
  // Navigation
  ('home', LucideIcons.home),
  ('menu', LucideIcons.menu),
  ('search', LucideIcons.search),
  ('settings', LucideIcons.settings),
  ('layoutDashboard', LucideIcons.layoutDashboard),
  ('layoutGrid', LucideIcons.layoutGrid),
  ('layoutList', LucideIcons.layoutList),
  ('list', LucideIcons.list),
  ('sidebar', LucideIcons.sidebar),
  ('compass', LucideIcons.compass),
  ('map', LucideIcons.map),
  ('mapPin', LucideIcons.mapPin),
  ('externalLink', LucideIcons.externalLink),
  ('link', LucideIcons.link),
  // People
  ('user', LucideIcons.user),
  ('userRound', LucideIcons.userRound),
  ('users', LucideIcons.users),
  ('usersRound', LucideIcons.usersRound),
  ('userPlus', LucideIcons.userPlus),
  ('userMinus', LucideIcons.userMinus),
  ('userCheck', LucideIcons.userCheck),
  ('smile', LucideIcons.smile),
  ('laugh', LucideIcons.laugh),
  ('frown', LucideIcons.frown),
  // Communication
  ('mail', LucideIcons.mail),
  ('mailOpen', LucideIcons.mailOpen),
  ('phone', LucideIcons.phone),
  ('phoneCall', LucideIcons.phoneCall),
  ('bell', LucideIcons.bell),
  ('bellRing', LucideIcons.bellRing),
  ('bellOff', LucideIcons.bellOff),
  ('messageSquare', LucideIcons.messageSquare),
  ('messageCircle', LucideIcons.messageCircle),
  ('atSign', LucideIcons.atSign),
  // Files & Folders
  ('file', LucideIcons.file),
  ('fileText', LucideIcons.fileText),
  ('fileCode', LucideIcons.fileCode),
  ('fileImage', LucideIcons.fileImage),
  ('filePlus', LucideIcons.filePlus),
  ('fileMinus', LucideIcons.fileMinus),
  ('folder', LucideIcons.folder),
  ('folderOpen', LucideIcons.folderOpen),
  ('folderPlus', LucideIcons.folderPlus),
  ('archive', LucideIcons.archive),
  ('package', LucideIcons.package),
  // Editing & Media
  ('image', LucideIcons.image),
  ('camera', LucideIcons.camera),
  ('video', LucideIcons.video),
  ('play', LucideIcons.play),
  ('pause', LucideIcons.pause),
  ('mic', LucideIcons.mic),
  ('volume', LucideIcons.volume),
  ('headset', LucideIcons.headset),
  ('bold', LucideIcons.bold),
  ('italic', LucideIcons.italic),
  ('heading', LucideIcons.heading),
  ('type', LucideIcons.type),
  // Status
  ('info', LucideIcons.info),
  ('circleCheck', LucideIcons.circleCheck),
  ('circleX', LucideIcons.circleX),
  ('circleAlert', LucideIcons.circleAlert),
  ('circleHelp', LucideIcons.circleHelp),
  ('alertTriangle', LucideIcons.alertTriangle),
  ('alertCircle', LucideIcons.alertCircle),
  ('shieldCheck', LucideIcons.shieldCheck),
  ('shieldAlert', LucideIcons.shieldAlert),
  ('zap', LucideIcons.zap),
  ('sparkles', LucideIcons.sparkles),
  ('flame', LucideIcons.flame),
  // Theme & Settings
  ('sun', LucideIcons.sun),
  ('moon', LucideIcons.moon),
  ('sunMoon', LucideIcons.sunMoon),
  ('eye', LucideIcons.eye),
  ('eyeOff', LucideIcons.eyeOff),
  ('lock', LucideIcons.lock),
  ('unlock', LucideIcons.unlock),
  ('key', LucideIcons.key),
  ('palette', LucideIcons.palette),
  ('layers', LucideIcons.layers),
  ('sliders', LucideIcons.sliders),
  // Time
  ('calendar', LucideIcons.calendar),
  ('clock', LucideIcons.clock),
  ('hourglass', LucideIcons.hourglass),
  ('timer', LucideIcons.timer),
  // Tech
  ('code', LucideIcons.code),
  ('terminal', LucideIcons.terminal),
  ('database', LucideIcons.database),
  ('server', LucideIcons.server),
  ('cloud', LucideIcons.cloud),
  ('cpu', LucideIcons.cpu),
  ('monitor', LucideIcons.monitor),
  ('laptop', LucideIcons.laptop),
  ('wifi', LucideIcons.wifi),
  ('bluetooth', LucideIcons.bluetooth),
  ('gitBranch', LucideIcons.gitBranch),
  ('gitCommit', LucideIcons.gitCommit),
  ('gitMerge', LucideIcons.gitMerge),
  ('gitPullRequest', LucideIcons.gitPullRequest),
];

class IconsPage extends StatefulWidget {
  const IconsPage({super.key});

  @override
  State<IconsPage> createState() => _IconsPageState();
}

class _IconsPageState extends State<IconsPage> {
  String _query = '';

  List<(String, IconData)> get _filtered {
    if (_query.isEmpty) return _icons;
    final q = _query.toLowerCase();
    return _icons.where((e) => e.$1.toLowerCase().contains(q)).toList();
  }

  @override
  Widget build(BuildContext context) {
    final c = SomaTheme.of(context);
    final filtered = _filtered;

    return Column(
      children: [
        // Search bar
        Padding(
          padding: const EdgeInsets.all(24),
          child: Row(
            children: [
              Expanded(
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    Text('Icons',
                        style: Theme.of(context).textTheme.headlineMedium),
                    const SizedBox(height: 4),
                    Text(
                      'Lucide icon set (parity with the web crate).',
                      style: TextStyle(
                        fontFamily: 'Outfit',
                        fontSize: 14,
                        color: c.mutedForeground,
                      ),
                    ),
                  ],
                ),
              ),
            ],
          ),
        ),
        Padding(
          padding: const EdgeInsets.fromLTRB(24, 0, 24, 16),
          child: Row(
            children: [
              Icon(LucideIcons.search, size: 16, color: c.mutedForeground),
              const SizedBox(width: 8),
              Expanded(
                child: SomaInput(
                  placeholder: 'Search icons…',
                  onChanged: (v) => setState(() => _query = v),
                ),
              ),
              const SizedBox(width: 12),
              Text(
                '${filtered.length} / ${_icons.length}',
                style: TextStyle(
                  fontFamily: 'Outfit',
                  fontSize: 12,
                  color: c.mutedForeground,
                ),
              ),
            ],
          ),
        ),

        // Grid
        Expanded(
          child: LayoutBuilder(builder: (context, constraints) {
            final cols = (constraints.maxWidth / 96).floor().clamp(3, 8);
            return filtered.isEmpty
                ? Center(
                    child: Text(
                      'No icons match "$_query"',
                      style: TextStyle(
                        fontFamily: 'Outfit',
                        fontSize: 14,
                        color: c.mutedForeground,
                      ),
                    ),
                  )
                : GridView.builder(
                    padding: const EdgeInsets.fromLTRB(24, 0, 24, 24),
                    gridDelegate: SliverGridDelegateWithFixedCrossAxisCount(
                      crossAxisCount: cols,
                      mainAxisSpacing: 8,
                      crossAxisSpacing: 8,
                      mainAxisExtent: 80,
                    ),
                    itemCount: filtered.length,
                    itemBuilder: (context, i) {
                      final (name, data) = filtered[i];
                      return _IconCell(name: name, data: data);
                    },
                  );
          }),
        ),
      ],
    );
  }
}

class _IconCell extends StatefulWidget {
  final String name;
  final IconData data;

  const _IconCell({required this.name, required this.data});

  @override
  State<_IconCell> createState() => _IconCellState();
}

class _IconCellState extends State<_IconCell> {
  bool _hovered = false;

  @override
  Widget build(BuildContext context) {
    final c = SomaTheme.of(context);
    return MouseRegion(
      cursor: SystemMouseCursors.click,
      onEnter: (_) => setState(() => _hovered = true),
      onExit: (_) => setState(() => _hovered = false),
      child: AnimatedContainer(
        duration: const Duration(milliseconds: 120),
        decoration: BoxDecoration(
          color: _hovered ? c.accent : Colors.transparent,
          border: Border.all(color: _hovered ? c.ring : c.border),
          borderRadius: BorderRadius.circular(6),
        ),
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            Icon(widget.data, size: 24, color: c.foreground),
            const SizedBox(height: 6),
            Padding(
              padding: const EdgeInsets.symmetric(horizontal: 4),
              child: Text(
                widget.name,
                textAlign: TextAlign.center,
                maxLines: 2,
                overflow: TextOverflow.ellipsis,
                style: TextStyle(
                  fontFamily: 'Outfit',
                  fontSize: 11,
                  color: c.mutedForeground,
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }
}

File: soma-ui/web/playground/src/pages/icons_page.rs
use crate::i18n::{strings, Locale};
use leptos::prelude::*;
use soma_ui::icons::{icondata, Icon};

// ponytail: curated ~120-icon subset of the full Lucide catalog (1 600+ icons).
// Ceiling: only what's listed here appears in the gallery.
// Upgrade path: add more (&str, icondata::Icon) pairs, or codegen the full
// list from `icondata_lu` at build time.
fn all_icons() -> Vec<(&'static str, icondata::Icon)> {
    vec![
        // Arrows
        ("ArrowUp", icondata::LuArrowUp),
        ("ArrowDown", icondata::LuArrowDown),
        ("ArrowLeft", icondata::LuArrowLeft),
        ("ArrowRight", icondata::LuArrowRight),
        ("ArrowUpLeft", icondata::LuArrowUpLeft),
        ("ArrowUpRight", icondata::LuArrowUpRight),
        ("ArrowDownLeft", icondata::LuArrowDownLeft),
        ("ArrowDownRight", icondata::LuArrowDownRight),
        ("ArrowUpDown", icondata::LuArrowUpDown),
        ("ArrowLeftRight", icondata::LuArrowLeftRight),
        // Chevrons
        ("ChevronUp", icondata::LuChevronUp),
        ("ChevronDown", icondata::LuChevronDown),
        ("ChevronLeft", icondata::LuChevronLeft),
        ("ChevronRight", icondata::LuChevronRight),
        ("ChevronsUp", icondata::LuChevronsUp),
        ("ChevronsDown", icondata::LuChevronsDown),
        ("ChevronsLeft", icondata::LuChevronsLeft),
        ("ChevronsRight", icondata::LuChevronsRight),
        // Actions / UI
        ("Check", icondata::LuCheck),
        ("X", icondata::LuX),
        ("Plus", icondata::LuPlus),
        ("Minus", icondata::LuMinus),
        ("Search", icondata::LuSearch),
        ("Menu", icondata::LuMenu),
        ("Grip", icondata::LuGrip),
        ("Ellipsis", icondata::LuEllipsis),
        ("EllipsisVertical", icondata::LuEllipsisVertical),
        ("RefreshCw", icondata::LuRefreshCw),
        ("RotateCw", icondata::LuRotateCw),
        ("RotateCcw", icondata::LuRotateCcw),
        ("ExternalLink", icondata::LuExternalLink),
        ("Link", icondata::LuLink),
        ("Share", icondata::LuShare),
        ("Copy", icondata::LuCopy),
        ("Clipboard", icondata::LuClipboard),
        ("ClipboardCheck", icondata::LuClipboardCheck),
        ("Download", icondata::LuDownload),
        ("Upload", icondata::LuUpload),
        ("Send", icondata::LuSend),
        ("ListFilter", icondata::LuListFilter),
        ("Maximize", icondata::LuMaximize),
        ("Minimize", icondata::LuMinimize),
        // Navigation / Layout
        ("Home", icondata::LuHouse),
        ("LayoutDashboard", icondata::LuLayoutDashboard),
        ("PanelLeft", icondata::LuPanelLeft),
        ("PanelRight", icondata::LuPanelRight),
        ("List", icondata::LuList),
        ("Grid2x2", icondata::LuGrid2x2),
        ("Grid3x3", icondata::LuGrid3x3),
        ("Columns2", icondata::LuColumns2),
        ("MoveHorizontal", icondata::LuMoveHorizontal),
        // People
        ("User", icondata::LuUser),
        ("Users", icondata::LuUsers),
        // Communication
        ("Mail", icondata::LuMail),
        ("Phone", icondata::LuPhone),
        ("Bell", icondata::LuBell),
        ("BellOff", icondata::LuBellOff),
        ("Globe", icondata::LuGlobe),
        // Files & Folders
        ("File", icondata::LuFile),
        ("FileText", icondata::LuFileText),
        ("FileCode", icondata::LuFileCode),
        ("FilePlus", icondata::LuFilePlus),
        ("FileMinus", icondata::LuFileMinus),
        ("FilePen", icondata::LuFilePen),
        ("Folder", icondata::LuFolder),
        ("Package", icondata::LuPackage),
        ("Save", icondata::LuSave),
        ("Archive", icondata::LuArchive),
        ("Book", icondata::LuBook),
        ("BookOpen", icondata::LuBookOpen),
        ("Bookmark", icondata::LuBookmark),
        ("Tag", icondata::LuTag),
        // Editing
        ("Pencil", icondata::LuPencil),
        ("Pen", icondata::LuPen),
        ("Trash", icondata::LuTrash),
        ("Trash2", icondata::LuTrash2),
        ("Scissors", icondata::LuScissors),
        // Media
        ("Image", icondata::LuImage),
        ("Camera", icondata::LuCamera),
        ("Video", icondata::LuVideo),
        ("Mic", icondata::LuMic),
        ("Play", icondata::LuPlay),
        ("Pause", icondata::LuPause),
        ("Volume", icondata::LuVolume),
        ("Volume2", icondata::LuVolume2),
        ("VolumeX", icondata::LuVolumeX),
        // Status / Feedback
        ("Star", icondata::LuStar),
        ("Heart", icondata::LuHeart),
        ("Flag", icondata::LuFlag),
        ("Zap", icondata::LuZap),
        ("Info", icondata::LuInfo),
        ("Circle", icondata::LuCircle),
        ("Square", icondata::LuSquare),
        ("Loader", icondata::LuLoader),
        ("Shield", icondata::LuShield),
        ("ShieldCheck", icondata::LuShieldCheck),
        // Theme / Settings
        ("Sun", icondata::LuSun),
        ("Moon", icondata::LuMoon),
        ("Cloud", icondata::LuCloud),
        ("Settings", icondata::LuSettings),
        ("SlidersHorizontal", icondata::LuSlidersHorizontal),
        ("Command", icondata::LuCommand),
        ("ToggleLeft", icondata::LuToggleLeft),
        ("ToggleRight", icondata::LuToggleRight),
        ("Eye", icondata::LuEye),
        ("EyeOff", icondata::LuEyeOff),
        ("Lock", icondata::LuLock),
        ("LockOpen", icondata::LuLockOpen),
        ("Key", icondata::LuKey),
        ("LogIn", icondata::LuLogIn),
        ("LogOut", icondata::LuLogOut),
        // Time & Calendar
        ("Calendar", icondata::LuCalendar),
        ("Clock", icondata::LuClock),
        ("Map", icondata::LuMap),
        // Tech
        ("Code", icondata::LuCode),
        ("Terminal", icondata::LuTerminal),
        ("Database", icondata::LuDatabase),
        ("Server", icondata::LuServer),
        ("Cpu", icondata::LuCpu),
        ("Monitor", icondata::LuMonitor),
        ("Laptop", icondata::LuLaptop),
        ("Smartphone", icondata::LuSmartphone),
        ("Tablet", icondata::LuTablet),
        ("HardDrive", icondata::LuHardDrive),
        ("Wifi", icondata::LuWifi),
        ("Bluetooth", icondata::LuBluetooth),
        // Social
        ("Github", icondata::LuGithub),
        ("Twitter", icondata::LuTwitter),
        ("Linkedin", icondata::LuLinkedin),
        ("Instagram", icondata::LuInstagram),
    ]
}

#[component]
pub fn IconsPage() -> impl IntoView {
    let locale = use_context::<RwSignal<Locale>>().expect("locale context");
    let s = move || strings(locale.get());
    let query = RwSignal::new(String::new());

    let icons = all_icons();

    let filtered = move || {
        let q = query.get().to_lowercase();
        icons
            .iter()
            .filter(|(name, _)| q.is_empty() || name.to_lowercase().contains(&q))
            .cloned()
            .collect::<Vec<_>>()
    };

    view! {
        <div class="max-w-5xl space-y-8">
            <div>
                <h1 class="font-heading text-3xl font-bold tracking-tight text-foreground">{move || s().icons_title}</h1>
                <p class="text-sm text-muted-foreground mt-1">{move || s().icons_subtitle}</p>
            </div>

            // Search
            <div class="bg-card border border-border rounded-md p-4 flex items-center gap-3">
                <Icon icon=Signal::derive(|| icondata::LuSearch) width="16" height="16" style="color: var(--color-muted-foreground)" />
                <input
                    type="text"
                    class="flex-1 bg-transparent text-sm text-foreground placeholder:text-muted-foreground outline-none"
                    placeholder=move || s().icons_search_placeholder
                    on:input=move |e| query.set(event_target_value(&e))
                />
            </div>

            // Grid
            <div class="bg-card border border-border rounded-md p-6">
                <div class="grid grid-cols-3 sm:grid-cols-4 md:grid-cols-6 lg:grid-cols-8 gap-4">
                    {move || {
                        let items = filtered();
                        if items.is_empty() {
                            view! {
                                <p class="col-span-full text-sm text-muted-foreground text-center py-8">"No icons match your search."</p>
                            }.into_any()
                        } else {
                            items.into_iter().map(|(name, icon)| {
                                view! {
                                    <div class="flex flex-col items-center gap-2 p-3 rounded-md hover:bg-accent transition-colors">
                                        <Icon
                                            icon=Signal::derive(move || icon)
                                            width="24"
                                            height="24"
                                            style="color: var(--color-foreground)"
                                        />
                                        <span class="text-xs text-muted-foreground text-center leading-tight break-all">{name}</span>
                                    </div>
                                }
                            }).collect::<Vec<_>>().into_any()
                        }
                    }}
                </div>
            </div>
        </div>
    }
}
